### DISCUSSION TITLE
Emitting `MaterializeResult` whilst returning a value

### DISCUSSION THREAD

**Discussion by christeefy** [dagster-io/dagster#26198] [Q&A] [answered]
## Context 
<details>
<summary> Context </summary>

Historically Dagster allows returning assets directly in its definition:
```python
@asset
def my_asset():
    return "Hello, world!"
```

although in the past year the above pattern is less recommended, in-lieu of each asset responsible for serialization & deserialization:
```python
@asset
def table():
    df = ...
    df.to_csv("...")

@asset(deps=[table])
def downstream1():
     df = pd.read_csv("...")
     ...
```

If I were to guess, this is to prevent anti-patterns like loading a massive table from the warehouse and passing it between assets. That said, it feels redundant to have each dependent asset worry about (de)serialization and knowledge of that upstream asset's internals. 

In any case, I see that `MaterializeResult` was designed to enable the second pattern above, as it doesn't have something like a `value` field when `yield`ed/`return`ed from the asset. 

</details>

## Problem
In the process of onboarding team members onto Dagster, we ran into the following use case (simplified):
```python
@asset
def latest_data_load_date():
    latest_date = "..."
    yield MaterializeResult(metadata={"latest_data_load_date": MetadataValue.timestamp(latest_date)})
    return latest_date

@asset
def downstream(latest_data_load_date: str):
    ...
```
We have two intentions with `latest_data_load_date`:
- It is to be used by multiple assets, so we want to centralize its definition in one place (as opposed to repeating it in multiple assets). 
- Display the latest load date in the Dagster UI to easily observe its value, and its history of values

The `@asset` API suggests that it's possible to both return data and `MaterializeResult` (see Context section). However, the code above fails with the following message:
<details>
<summary>Error stack trace</summary>

```
dagster._core.errors.DagsterExecutionLoadInputError: Error occurred while loading input "latest_data_load_date" of step "downstream":

  File "/Users/<usr>/Projects/gdp-dagster/.venv/lib/python3.11/site-packages/dagster/_core/execution/plan/execute_plan.py", line 245, in dagster_event_sequence_for_step
    for step_event in check.generator(step_events):
  File "/Users/<usr>/Projects/gdp-dagster/.venv/lib/python3.11/site-packages/dagster/_core/execution/plan/execute_step.py", line 465, in core_dagster_event_sequence_for_step
    for event_or_input_value in step_input.source.load_input_object(
  File "/Users/<usr>/Projects/gdp-dagster/.venv/lib/python3.11/site-packages/dagster/_core/execution/plan/inputs.py", line 363, in load_input_object
    yield from _load_input_with_input_manager(input_manager, load_input_context)
  File "/Users/<usr>/Projects/gdp-dagster/.venv/lib/python3.11/site-packages/dagster/_core/execution/plan/inputs.py", line 613, in _load_input_with_input_manager
    with op_execution_error_boundary(
  File "/Users/<usr>/.local/share/uv/python/cpython-3.11.10-macos-aarch64-none/lib/python3.11/contextlib.py", line 158, in __exit__
    self.gen.throw(typ, value, traceback)
  File "/Users/<usr>/Projects/gdp-dagster/.venv/lib/python3.11/site-packages/dagster/_core/execution/plan/utils.py", line 84, in op_execution_error_boundary
    raise error_cls(

The above exception was caused by the following exception:
FileNotFoundError: [Errno 2] No such file or directory: '/Users/<usr>/Projects/gdp-dagster/tmpgdig4yno/storage/latest_data_load_date'

  File "/Users/<usr>/Projects/gdp-dagster/.venv/lib/python3.11/site-packages/dagster/_core/execution/plan/utils.py", line 54, in op_execution_error_boundary
    yield
  File "/Users/<usr>/Projects/gdp-dagster/.venv/lib/python3.11/site-packages/dagster/_core/execution/plan/inputs.py", line 620, in _load_input_with_input_manager
    value = input_manager.load_input(context)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/<usr>/Projects/gdp-dagster/.venv/lib/python3.11/site-packages/dagster/_core/storage/upath_io_manager.py", line 405, in load_input
    return self._load_single_input(path, context)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/<usr>/Projects/gdp-dagster/.venv/lib/python3.11/site-packages/dagster/_core/storage/upath_io_manager.py", line 272, in _load_single_input
    obj = self.load_from_path(context=context, path=path)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/<usr>/Projects/gdp-dagster/.venv/lib/python3.11/site-packages/dagster/_core/storage/fs_io_manager.py", line 283, in load_from_path
    with path.open("rb") as file:
         ^^^^^^^^^^^^^^^
  File "/Users/<usr>/Projects/gdp-dagster/.venv/lib/python3.11/site-packages/upath/implementations/local.py", line 134, in open
    return PosixPath.open(self, mode, buffering, encoding, errors, newline)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/<usr>/.local/share/uv/python/cpython-3.11.10-macos-aarch64-none/lib/python3.11/pathlib.py", line 1044, in open
    return io.open(self, mode, buffering, encoding, errors, newline)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```
</details>

Further investigation show that the plumbing between assets are broken with the use of `MaterializeResult` in tandem with returning the asset. 

And so, we have to choose between fulfilling _either_ of our requirements for the asset, but not both. 

So the asks are:
- Is there a way to achieve both requirements? Are we abusing the use of `MaterializeResult` instead of using `context.log`? 
- The error message above can be improved upon (happy to contribute with guidance!)

[Other](https://dagster.slack.com/archives/C066HKS7EG1/p1729390697414639) [users](https://dagster.slack.com/archives/C066HKS7EG1/p1727263790031979) also have the same question, so it would be nice to gain clarification on proper usage / composition of these APIs. 

Thanks!

**Comment by alangenfeld** [comment, marked as answer]
You can switch from `MaterializeResult` to `Output` when you want to return a value in addition to `metadata` 

```
@asset
def latest_data_load_date():
    latest_date = "..."
    return Output(latest_date, metadata={"latest_data_load_date": MetadataValue.timestamp(latest_date)})
```

will follow up on the right place to put this in the docs

**Reply by v1gnesh** [reply]
Also, can you please point me/us to docs on why pattern 2 was chosen in favour of pattern 1. Is it solely because of the massive table passing mentioned above or... more reasons? From my perspective, I see the benefit of both - each asset's effect is separate, it makes debugging easier if there's a usable file on disk, rather than a pickle. On the other hand, maybe for simpler use cases, writing and reading back with pandas gets added everywhere. On that topic, please consider pointing to polars instead of pandas... or at least specify both. Polars deserves to supercede pandas, and it does today, at least for a fair portion of use cases.

**Reply by alangenfeld** [reply]
"pattern 1" represents Dagster managed inputs/outputs which are handled by a subsystem called IO Managers https://docs.dagster.io/concepts/io-management/io-managers

We have found over the years that IO Managers often create a rough learning curve for new users. When using this pattern, new users are able to move values from asset to asset without even knowing what an IO manager is or how it works. Then at some point the default IO manager stops working and the user has to learn a whole bunch of new concepts at a time when they just want to get things running again. 

For this reason, we have started removing managed IO from the examples in documentation that we expect new users to encounter early on as they learn how to use Dagster. The goal is to make it so that users are introduced to it a more appropriate time, when they are better equipped to understand its value and which IO manager they may need for different assets and different deployments/environments. This is something we are still working on, you can give feedback on the new docs revamp we are working on here https://github.com/dagster-io/dagster/discussions/24816

**Reply by christeefy** [reply]
Thanks @alangenfeld for the context. I found this super helpful to know that Pattern 1 is not part of a deprecated feature, but an advanced one. 

I think it may also be valuable to include the context you shared above somewhere in the documentation. It may be helpful for other users facing the same question.
